### PR TITLE
Add length-unbiased GRPO loss (LUSPO)

### PR DESCRIPTION
**Pull request by Haseebasif7**
# What does this PR do?

This PR implements the **Length-Unbiased Sequence Policy Optimization (LUSPO)** loss in the [GRPOTrainer](cci:2://file:///Users/haseeb/Desktop/trl/trl/trainer/grpo_trainer.py:121:0-2215:46). 

### Motivation
Standard GRPO involves sequence-level normalization that can sometimes introduce length bias (preferring shorter completions with positive rewards or longer ones with negative rewards). LUSPO addresses this by aggregating token-level losses differently: it computes the sum of per token losses (using standard PPO style clipping) for each sequence and then averages these sequence-level sums across the batch, effectively removing the per sequence length normalizer found in the base "grpo" loss type.

### Changes
- **[GRPOTrainer](cci:2://file:///Users/haseeb/Desktop/trl/trl/trainer/grpo_trainer.py:121:0-2215:46)**: Added `luspo` case in [_compute_loss](cci:1://file:///Users/haseeb/Desktop/trl/trl/trainer/grpo_trainer.py:1944:4-2129:19) for both token-level loss calculation and batch aggregation.
- **[GRPOConfig](cci:2://file:///Users/haseeb/Desktop/trl/trl/trainer/grpo_config.py:21:0-901:86)**: Updated [loss_type](cci:1://file:///Users/haseeb/Desktop/trl/tests/test_grpo_trainer.py:207:4-237:87) documentation to include `luspo`.
- **Tests**: Added `luspo` to the automated loss type test suite in [tests/test_grpo_trainer.py](cci:7://file:///Users/haseeb/Desktop/trl/tests/test_grpo_trainer.py:0:0-0:0).
 
**Reference:**
- **LUSPO Paper:** [Length-Unbiased Sequence Policy Optimization](https://arxiv.org/abs/2602.05261)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?
@noahlt @kashif @namin

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -2079,6 +2079,11 @@ def _compute_loss(self, model, inputs):
         elif self.loss_type in ["cispo", "dapo"]:
             normalizer = inputs["num_items_in_batch"] / self.accelerator.num_processes
             loss = (per_token_loss * mask).sum() / normalizer
+        elif self.loss_type == "luspo":
+            # Length-Unbiased Sequence Policy Optimization (LUSPO)
+            loss = (per_token_loss * mask).sum(-1).mean()
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]:
```suggestion
            loss = (per_token_loss * mask.sum(-1)).mean()
```
Mirrors the paper's $|y_i|$ scaling (eq. 7) directly. The alternative `(... * mask).sum(-1)` is equivalent but requires the reader to reason about why the per-response normalization is absent.

File: trl/trainer/grpo_config.py
Comment (LeonEricsson) [inline trl/trainer/grpo_config.py:1]: Need to add 'luspo' documentation in the GRPOConfig docstring. Line ~239

File: trl/trainer/grpo_trainer.py
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]:
```suggestion
        if self.loss_type in ["grpo", "bnpo", "dr_grpo", "dapo", "luspo"]:
```

**Comment by Haseebasif7** [comment]
> Looks good, a few nits. Could you add a Paper Index as well?
> 
> Unlike GSPO, which doesn't have it's own loss_type, LUSPO is a modification to GSPO, meaning it is dependent on `importance_sampling_level=sequence". We should make this clear both in the documentation in grpo_config, as well as in the paper index.


Thanks for the Review , I had done the required changes

File: docs/source/paper_index.md [hunk 2]
```diff
@@ -413,6 +413,22 @@ training_args = GRPOConfig(
 )
 ```
 
+### Length-Unbiased Sequence Policy Optimization
+
+**📜 Paper**: https://arxiv.org/abs/2602.05261
```
Comment (qgallouedec) [inline docs/source/paper_index.md]:
```suggestion
**📜 Paper**: https://huggingface.co/papers/2602.05261
```

File: docs/source/paper_index.md [hunk 3]
```diff
@@ -413,6 +413,22 @@ training_args = GRPOConfig(
 )
 ```
 
+### Length-Unbiased Sequence Policy Optimization
```
Comment (qgallouedec) [inline docs/source/paper_index.md]:
```suggestion
### Length-Unbiased Sequence Policy Optimization: Revealing and Controlling Response Length Variation in RLVR
```

also please move this section just below GDPO? to keep a chronological order inside subsection

File: trl/trainer/grpo_config.py [hunk 4]
```diff
@@ -233,10 +233,10 @@ class GRPOConfig(TrainingArguments):
               clipped weights are then multiplied with the advantages and policy model's log probs. Individual token
               losses are aggregated by normalizing with the number of active tokens in the global accumulated batch.
               This method was introduced in the [MiniMax-M1 paper](https://huggingface.co/papers/2506.13585).
-            - `"sapo"`: Soft Adaptive Policy Optimization loss, as introduced in the [Soft Adaptive Policy Optimization
-              paper](https://huggingface.co/papers/2506.13585). Replaces hard clipping with a smooth,
-              temperature-controlled gate that adaptively attenuates off-policy updates while preserving useful
-              learning signals.
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py:237]: please revert this deletion

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -2079,6 +2082,11 @@ def _compute_loss(self, model, inputs):
         elif self.loss_type in ["cispo", "dapo"]:
             normalizer = inputs["num_items_in_batch"] / self.accelerator.num_processes
             loss = (per_token_loss * mask).sum() / normalizer
+        elif self.loss_type == "luspo":
+            # Length-Unbiased Sequence Policy Optimization (LUSPO)
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]:
```suggestion
```

File: trl/trainer/grpo_trainer.py [hunk 6]
```diff
@@ -2079,6 +2082,11 @@ def _compute_loss(self, model, inputs):
...
+            loss = (per_token_loss * mask.sum(-1)).mean()
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]: 👌

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: are you sure about this one?

```python
loss = (per_token_loss * mask.sum(-1)).mean()
```

broadcasts `mask.sum(-1)` over tokens, so it scales every token by seq length, ie each sequence gets weighted by |y|^2, not by **|y|**.

File: trl/trainer/grpo_trainer.py (hunk 6 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: actually, maybe not, see https://github.com/huggingface/trl/pull/4988#discussion_r2775943589

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py]: sorry, the correct formulation is:

`loss = (per_token_loss.squeeze(1) * mask.sum(-1).mean()`
or 
`loss = (per_token_loss * mask.sum(-1, keepdim=True)).mean()`

otherwise we get an outer-product-like result [B, B]. but I don't think that's the error you were implying @qgallouedec? note that per_token_loss is a sequence loss [B, 1], so there are no tokens to broadcast over

Comment (Haseebasif7) [reply trl/trainer/grpo_trainer.py]: updated

**Comment by Haseebasif7** [comment]
@qgallouedec @LeonEricsson I had applied the required changes

**Comment by LeonEricsson** [comment]
> I had applied the required changes

Could you resolve the conflict with test_grpo_trainer.py on main while we wait for Quentin's approval, I want to see what he says about [this](https://github.com/huggingface/trl/pull/4988#discussion_r2776457146)

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: yep that's what I meant Leon, let me check the new form

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: Ok it's correct, sorry for the back and forth, but I feel like 

```python
loss = (per_token_loss.squeeze(1) * mask.sum(-1)).mean()
```

is actually more aligned with the paper: it uses 1D tensor, which is align with the single indexing in equation (7)

I will push this and add a small comment explaining this

> note that per_token_loss is a sequence loss [B, 1], so there are no tokens to broadcast over

**Comment by Haseebasif7** [comment]
> > I had applied the required changes
> 
> Could you resolve the conflict with test_grpo_trainer.py on main while we wait for Quentin's approval, I want to see what he says about [this](https://github.com/huggingface/trl/pull/4988#discussion_r2776457146)

got it !

File: trl/trainer/grpo_trainer.py [hunk 7]
```diff
@@ -517,6 +517,9 @@ def __init__(
                 "When using `sapo` loss, both `sapo_temperature_neg` and `sapo_temperature_pos` must be set."
             )
 
+        if args.loss_type == "luspo" and args.importance_sampling_level != "sequence":
+            raise ValueError("When using `luspo` loss, `importance_sampling_level` must be set to `sequence`.")
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]: I think a warning is sufficient. The docs are very clear that importance_sampling_level _should_ be sequence, according to the paper.

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: yep agree lemme update this
